src/s390x/dasd.rs

```diff
@@ -0,0 +1,284 @@
+    if !status.success() {
+        return Err("Execution failed".into());
+    }
+    match status.code() {
+        Some(code) => {
+            if code != 0 {
+                return Err(format!("Failed with code: {}", code).into());
+            }
+        }
+        None => {
+            return Err(("Process terminated by signal").into());
+        }
+    }
+    Ok(())
+}
+
+/// Download and unpack CoreOS image to file.
+/// Return path to raw image or error
+///
+/// # Arguments
+/// * `source` - CoreOS image source
+fn save_image(source: &mut ImageSource) -> Result<PathBuf> {
+    let mut image_filename = PathBuf::new();
+    image_filename.push("/tmp/");
+    image_filename.push(
+        source
+            .filename
+            .trim_end_matches(".gz")
+            .trim_end_matches(".xz"),
+    );
+    eprintln!("Writing CoreOS image to {}", image_filename.display());
+    let mut image = File::create(&image_filename)
+        .chain_err(|| format!("creating {}", image_filename.display()))?;
+    write_image(source, &mut image, true)?;
+    Ok(image_filename)
+}
+
+/// Generates options for creating and copying disk partitions
+///
+/// # Arguments
+/// * `dasd` - dasd device, i.e. smth like /dev/dasda
+/// * `image` - CoreOS raw disk image
+fn partitioning_options<P1: AsRef<Path>>(
+    dasd: &str,
+    image: P1,
+) -> Result<(Vec<Opts>, Vec<String>)> {
+    eprintln!("Generating partitioning options");
+    let (blocks, bytes) = disk_config(dasd)?;
+    let image = image.as_ref();
+    let mut ctx = Context::new();
+    ctx.save_user_sector_size(bytes as u32, bytes as u32)
+        .chain_err(|| format!("setting block size {}", bytes))?;
+    ctx.assign_device(&image, true)
+        .chain_err(|| format!("opening image {}", image.display()))?;
+
+    let mut dd_options = Vec::new();
+    let mut partitions = Vec::new();
+    let mut start = 2; //the first 2 tracks of the ECKD DASD are reserved;
+    let mut table = ctx.get_partitions().chain_err(|| "reading partitions")?;
+    let mut count = table.nents();
+
+    for mut pt in &mut table {
+        let uuid = pt.uuid().unwrap_or_else(|_| "UUID unknown".to_string());
+        let skip = pt.start().chain_err(|| format!("{}: no start", uuid))?;
+        let size = pt.size().chain_err(|| format!("{}: no size", uuid))?;
+        let seek = blocks * start;
+        let end = (size + blocks) / 12;
+        eprintln!("{} start: {} size: {}", uuid, skip, size);
+        let mut opts = Opts::default();
+        opts.input_block_size = bytes as usize;
+        opts.output_block_size = bytes as usize;
+        opts.iflags = dd_lib::opts::IFlag::FULLBLOCK;
+        opts.input_file = Some(image.to_str().unwrap().to_string());
+        opts.output_file = Some(dasd.to_string());
+        opts.input_seek = Some(skip as usize);
+        opts.count = Some(size as usize);
+        opts.output_seek = Some(seek as usize);
+        dd_options.push(opts);
```

Comment:
It looks like we're doing a straightforward copy, without any fancy conversions.  Do we need to pull in `dd_lib` rather than just doing the I/O ourselves?

Reply:
dd is very convenient and less err-prone  way of copying. that's why it used here.